subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/FileChange.java

```diff
@@ -18,22 +18,52 @@
 
 import com.google.common.base.Objects;
 import org.gradle.api.tasks.incremental.InputFileDetails;
+import org.gradle.internal.nativeintegration.filesystem.FileType;
 
 import java.io.File;
 
 public class FileChange implements TaskStateChange, InputFileDetails {
     private final String path;
     private final ChangeType change;
-    private final String fileType;
+    private final String title;
+    private final FileType previousFileType;
+    private final FileType currentFileType;
 
-    public FileChange(String path, ChangeType change, String fileType) {
+    public static FileChange added(String path, String title) {
+        return new FileChange(path, ChangeType.ADDED, title, FileType.Missing, FileType.RegularFile);
+    }
+
+    public static FileChange removed(String path, String title) {
+        return new FileChange(path, ChangeType.REMOVED, title, FileType.RegularFile, FileType.Missing);
+    }
```

Comment:
I feel bad about storing incorrect data here. Can we pass in the actual type instead of using `RegularFile` always?